### PR TITLE
feat(formations): invite partners to the formation from the sidebar

### LINKED ISSUE #2724: Formation checklist: list people on the formation and invite partners from the sidebar
## Summary

Add a **People on this formation** section to the right-hand rail of the formation checklist page (both `/project/formation` and the staff drill-down `/foundation/formations/:slug`), under the existing formation card.

- Readers (anyone who can load the checklist) see who is on the formation, grouped into **LF Staff** (`@linuxfoundation.org`) and **Invited** (everyone else), with avatar, name, job title / organization, and assigned-item count.
- External rows carry a status: **Invited** once the person has an LF account linked (settings entry has a `username`), **Invite Sent** while the invite is pending (email-only entry).
- Writers (`can_write`) get an **Invite** action: Name + Email + role (View by default, Manage optional). The invite is a project invite (per #2147): the BFF `POST /api/projects/:uid/permissions` add flow, directory lookup first, manual entry on a `NOT_FOUND` miss so upstream sends the invite email.

## Design

Screenshot from the design repo session (18 Sep 2026): "People on this formation" rail section with LF STAFF / INVITED groups, Invited / Invite pending labels, footer note "Invited people see this project and the items assigned to them. Labels are for display; access comes from grants."

## Backend notes

- Upstream `lfx-v2-project-service` sends the invite for any settings entry with an empty `username` and promotes the entry in place on acceptance (`docs/lfid-invite-flow.md`).
- Settings carry only name/email/username/avatar, so job title and organization are enriched best-effort from the auth-service user-metadata read.
- Settings GET is gated on the bare project `auditor` relation; when it is refused (global-grant staff) the rail degrades to a "People unavailable" state rather than failing the checklist.

## Delivery

Two PRs: (1) read-only people list (types, BFF endpoint, card, tests); (2) invite dialog and flow.

Refs #1965 (Epic 1).

### PR DESCRIPTION
**Pull request by niravpatel27**
## Summary

PR 2 of 2 for #2724, on top of the merged people list (#2725): writers get an **Invite** action on the "People on this formation" card.

- **Invite dialog**: Name, Email, and a View / Manage radio (View by default — everyone invited from the card is read-only unless the inviter says otherwise; Manage is offered because View cannot act on checklist items).
- **A formation invite is a project invite** (#2147). Submission goes to the existing `POST /api/projects/:uid/permissions` add flow, directory lookup first: an address with an LF account is added outright and shows as **Invited** immediately. On the BFF's `NOT_FOUND` directory miss the same address is re-sent with the name, which the BFF stores as an email-only entry — the shape `lfx-v2-project-service` emails an invite for (its [`docs/lfid-invite-flow.md`](https://github.com/linuxfoundation/lfx-v2-project-service/blob/main/docs/lfid-invite-flow.md), in the upstream repo) and promotes in place on acceptance. That row shows **Invite Sent** until then.
- An address already on the project is rejected inline with no request. On success the card evicts the shared settings cache (so the Permissions page agrees) and re-reads its list; failures surface as an error toast with the dialog left open.
- The dialog is opened through `DialogService`, like this module's staff-edit and reason-prompt dialogs (no inline `<p-dialog>`); it owns the two-step add, the toasts, and the settings-cache eviction, and closes with its outcome so the card re-reads its list. Invite is offered only once the list has loaded (before that the duplicate guard has nothing to check, and while the list is unavailable the add would fail the same way).
- No new endpoints and no server changes for the invite itself.

## Also included

The first commit (`2d81ed0c0`) is the last review fix from #2725, which landed two minutes after that PR merged: the user-metadata memo now caches only the three rendered fields (title, organization, picture), each runtime-checked, so the raw profile with address/phone/bio never enters the process-wide cache. Spec: *memoises only the three rendered fields, never the raw profile and its PII*.

## Testing

- Unit: `formation-invite-dialog.component.spec.ts` (validation, duplicate block, normalised emit, reset-on-open, in-flight guard), `formation-people-card.component.spec.ts` invite cases (reader/writer affordance, single-request add, directory-miss re-send with the Manage role, error toast, in-flight guard), `formation.service.spec.ts` PII projection.
- E2E: `formation-people.spec.ts` gains the invite scenarios (single-request add with list re-read, directory-miss re-send, Manage role, duplicate, empty/malformed input, failed add) and the reader case; `formation-people-robust.spec.ts` covers the dialog's testid contract. All 17 cases in the two specs pass locally on Chromium.
- `yarn build`, `yarn lint`, `yarn check-types`, `./check-headers.sh`, `check-fixture-emails.sh`, `yarn e2e:check-collection` all green.

## Review follow-ups (already in the branch)

- `a24842bc9`: the permissions add route now runs the strict writer gate before any directory lookup (closing the 404-vs-403 directory-membership oracle), refuses re-adding an existing member with 409 `ALREADY_ON_PROJECT` instead of silently re-filing their role, and masks identifiers in its logs.
- `3c1297fa8`: trims before validating, hides Invite while the list is loading or re-reading, caps the dialog at 90vw, never treats an empty email as a duplicate, uses `lfx-button`, binds the submit to the dialog's lifetime, caps the name at 200 characters.

## Notes for reviewers

- Branch name carries an `-invite` suffix because `feat/issue-2724` is the merged PR 1 branch for the same issue; the suffix disambiguates PR 2 of #2724.

- Invite email delivery depends on the project service's `invitesEnabled` value per environment; the UI shows "Invite Sent" from the settings entry regardless.
- Role change and removal stay on the Permissions page; this card only lists and invites.

Closes #2724

### CODE REVIEW

**Comment by dealako** [comment]
Hey @niravpatel27, thanks for the thorough PR description and test plan — made this much easier to review.

**Overall impression:** This is a well-scoped, well-tested feature that follows the module's existing dialog conventions closely, and the accompanying memoization fix (only caching the three rendered profile fields) is a solid, deliberate privacy improvement. The client-side UX (duplicate guard, in-flight guard, directory-miss handling) is thoughtfully built. The main gap is that the authorization and PII-safety guarantees for the new invite flow live entirely on the client/BFF-log side rather than being enforced server-side — the existing sibling route (`updateProjectStaff`) already solved the exact same authz-oracle problem this PR's endpoint reintroduces, so this looks like a gap rather than a hard problem.

**Blocking: 2 issues (incl. 1 privacy)**
- Missing server-side writer gate on `POST /api/projects/:uid/permissions` creates a directory-membership oracle: `apps/lfx-one/src/server/controllers/project.controller.ts:261-333` (route: `apps/lfx-one/src/server/routes/projects.route.ts:27`) — any authenticated user with mere read access can distinguish "address has an LF account" (404→then-403) from "no account" (404 NOT_FOUND), since the writer-only PUT gate runs *after* the directory lookup. The sibling `updateProjectStaff` (`project.service.ts:795-808`) already added `checkSingleAccessStrict(..., access: 'writer')` for this exact reason — this route needs the same gate.
- Invitee email logged unmasked in the BFF success log: `apps/lfx-one/src/server/controllers/project.controller.ts:332` — every sibling log call in the same chain masks this value (`maskIdentifierForLogs`/`maskEmailForLogs`, e.g. `project.service.ts:759-777`), and pino's redact list doesn't cover it either.

**Minor: 6 issues**
- Add-is-upsert can silently overwrite/demote an existing member's role: the duplicate guard matches on settings *email* (`formation-invite-dialog.component.ts:79`) while the server matches on resolved *username* (`project.service.ts:688,698-699,746-750`) — a username-only member (no email in settings) slips past the guard and gets silently re-added at the invite's role, demoting e.g. Manage → View with a success toast.
- That same email-less-member case also makes the duplicate guard false-positive on the *empty* Email field on dialog open (`formation-people-card.component.ts:88`, `formation-invite-dialog.component.ts:167-170`), permanently hiding the "Enter an email address." required-error for the whole project.
- Email is validated with `Validators.pattern` before trimming, so a pasted address with leading/trailing whitespace is rejected as "invalid" even though it's later trimmed anyway: `formation-invite-dialog.component.ts:62,98,103`.
- No test covers the miss-then-failure branch of the two-step add (directory miss, then the re-send also fails) — the riskiest path in the flow, and the one where `submitting` must reset correctly with a non-dismissable dialog (`closable: false`): `formation-invite-dialog.component.spec.ts:104-128`.
- The dialog hand-rolls Cancel/Submit `<button>` elements with custom Tailwind classes and a manual spinner instead of the `lfx-button` wrapper used by the two dialogs it says it mirrors (`staff-edit-dialog`, `reason-prompt-dialog`): `formation-invite-dialog.component.html:74-93`.
- The eviction-placement comment says "DialogService is root-scoped, so this dialog can outlive the card" — but the card provides `DialogService` at component scope (`formation-people-card.component.ts:56`), so the premise is inaccurate; the submit subscription also has no `takeUntilDestroyed`: `formation-invite-dialog.component.ts:124-126`.

**Nit: 1 issue**
- The invite `name` field is unbounded free text with no length cap, persisted and forwarded into an outbound invite email on the directory-miss path: `formation-invite-dialog.component.ts:329` (matches the pre-existing `staff-edit-dialog` precedent, so not new risk, but worth tightening here).

**Question: 1 item**
- The PR description points at `docs/lfid-invite-flow.md` for the invite-flow shape, but that file doesn't exist anywhere in the repo's history — is this meant to be added, or was the description referencing something from another repo?

**Privacy: 1 finding (blocking)**
(included in the Blocking count above)
1. Unmasked invitee email in BFF logs: `apps/lfx-one/src/server/controllers/project.controller.ts:332` logs the raw `identifier` (always the invitee's email for this feature) while every sibling call masks the same value — fix by wrapping in `maskIdentifierForLogs(identifier)` as `project.service.ts` does elsewhere in the same call chain.

**AI bot reconciliation:** Cursor Bugbot posted only a PR summary (no line-level findings) and CodeRabbit's auto-review is disabled on this repo (skipped) — nothing to agree or disagree with from either.

**Needs changes before approval.**

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by dealako — DISMISSED** [review]
Summary posted: https://github.com/linuxfoundation/lfx-self-serve/pull/2728#issuecomment-5734637581

Requesting changes for: a missing server-side writer gate on `POST /api/projects/:uid/permissions` that creates a directory-membership oracle, and an unmasked invitee email in the BFF success log (privacy, blocking). See the summary comment for the full findings list, including several minor correctness issues in the duplicate-guard/role-overwrite path and test-coverage gaps.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/lfx-one/src/app/modules/dashboards/components/formation-invite-dialog/formation-invite-dialog.component.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+  protected onSubmit(): void {
+    if (this.submitting()) {
+      return;
+    }
+
+    // Submit stays enabled so a click surfaces the validation messages; the guard is here.
+    this.form.markAllAsTouched();
+    if (this.form.invalid || this.isDuplicate()) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (niravpatel27) [reply apps/lfx-one/src/app/modules/dashboards/components/formation-invite-dialog/formation-invite-dialog.component.ts:111]: Fixed in 3c1297fa8de3fa4158fdf5de15aaf5442e371b32 — `onSubmit` now patches the trimmed name and email back into the form before `markAllAsTouched`/validation, so a pasted address with surrounding whitespace passes the anchored pattern and is lowercased on emit. Spec: *accepts a pasted address with surrounding whitespace and normalises it before validating*.

File: apps/lfx-one/src/app/modules/dashboards/components/formation-people-card/formation-people-card.component.ts [hunk 2]
```diff
@@ -51,19 +71,61 @@ export class FormationPeopleCardComponent {
...
+  /** Lowercased addresses already listed — handed to the dialog, which rejects them inline with no request. */
+  protected readonly existingEmails = computed(() => this.response().people.map((person) => person.email.trim().toLowerCase()));
+  /**
+   * Invite is offered only once the list has loaded: the seed value and the loading window are
+   * both the `unavailable` shape, so this one condition covers both. Before the list is in,
+   * `existingEmails` would be empty and a re-add of an existing writer as View would silently
+   * demote them; while the settings read is refused, the add's own settings read fails the same
+   * way, so the button would offer an action that can only end in the error toast.
+   */
+  protected readonly canInvite = computed(() => this.canWrite() && !this.unavailable());
```
[the remark this answers is not in the document]
Comment (niravpatel27) [reply apps/lfx-one/src/app/modules/dashboards/components/formation-people-card/formation-people-card.component.ts]: Fixed in 3c1297fa8de3fa4158fdf5de15aaf5442e371b32 — `canInvite` is now `canWrite && !unavailable && !loading`, so the button disappears for the re-read window after an invite (and before the first read), and the dialog can never open on a stale `existingEmails` list. Spec: *hides the Invite action while the list re-reads after an invite, so a stale duplicate list is never used*. Server-side, a24842bc92b7ea48c3d51f3542b60dbee679547d additionally makes `add` refuse an existing member with 409 `ALREADY_ON_PROJECT` instead of re-filing their role, so even a stale client list can no longer demote anyone.

File: apps/lfx-one/src/app/modules/dashboards/components/formation-people-card/formation-people-card.component.ts [hunk 3]
```diff
@@ -51,19 +71,61 @@ export class FormationPeopleCardComponent {
...
+  protected openInvite(): void {
+    if (!this.canInvite()) {
+      return;
+    }
+
+    const data: FormationInviteDialogData = { projectUid: this.projectUid(), existingEmails: this.existingEmails() };
+
+    const ref: DynamicDialogRef | null = this.dialogService.open(FormationInviteDialogComponent, {
+      header: FORMATION_INVITE_DIALOG_HEADER,
+      width: '500px',
```
[the remark this answers is not in the document]
Comment (niravpatel27) [reply apps/lfx-one/src/app/modules/dashboards/components/formation-people-card/formation-people-card.component.ts:113]: Fixed in 3c1297fa8de3fa4158fdf5de15aaf5442e371b32 — the `DialogService.open` config now passes `style: { maxWidth: '90vw' }` alongside `width: '500px'`, the same cap the repo's other dynamic dialogs use (e.g. `org-easycla-detail`, `account-settings`).

**Comment by niravpatel27** [comment]
Thanks @dealako — everything addressed across two commits.

**Blocking**
- **Writer gate / directory oracle** — a24842bc92b7ea48c3d51f3542b60dbee679547d: `updateProjectPermissions` now runs the same `checkSingleAccessStrict({ access: 'writer' })` gate as `updateProjectStaff` before the settings read and before any directory lookup, failing closed on an access-service error. Specs: *rejects a non-writer before the settings read or the directory lookup* (asserts no NATS, no settings read, no write) and *fails closed when the access check itself cannot be resolved*.
- **Unmasked invitee email** — a24842bc92b7ea48c3d51f3542b60dbee679547d: the add/update/remove permission routes now log `maskIdentifierForLogs(identifier)` in both `startOperation` and `success`, matching the service-layer logs in the same chain.

**Minor**
- **Add-is-upsert can demote** — a24842bc92b7ea48c3d51f3542b60dbee679547d: `add` now refuses anyone already listed (matched by username or, for email-only entries, by email) with 409 `ALREADY_ON_PROJECT` instead of re-filing them under the submitted role; `update` keeps re-filing roles. Spec: *refuses to re-add someone already on the project instead of silently re-filing their role*. The dialog surfaces the 409 through `serverAuthoredMessage`.
- **Empty-email false positive** — 3c1297fa8de3fa4158fdf5de15aaf5442e371b32: the card filters empty addresses out of `existingEmails`, and the dialog never treats an empty field as a duplicate, so the required error always shows. Specs on both sides.
- **Trim before validate** — 3c1297fa8de3fa4158fdf5de15aaf5442e371b32, as above.
- **Miss-then-failure test** — 3c1297fa8de3fa4158fdf5de15aaf5442e371b32: *surfaces a failure of the re-send after a directory miss as an error toast, resets submitting, and stays open*.
- **lfx-button** — 3c1297fa8de3fa4158fdf5de15aaf5442e371b32: Cancel and Send invite are `lfx-button`s with `[loading]`, as in `staff-edit-dialog`.
- **Eviction comment / takeUntilDestroyed** — 3c1297fa8de3fa4158fdf5de15aaf5442e371b32: comment reworded to the real reason (evict where the write happened so the shared settings cache can't serve the pre-add document, whichever host opened the dialog); the submit subscription is now `takeUntilDestroyed`.

**Nit** — 3c1297fa8de3fa4158fdf5de15aaf5442e371b32: name capped at `FORMATION_INVITE_NAME_MAX_LENGTH` (200) with both a validator and the input's `maxlength`.

**Question** — `docs/lfid-invite-flow.md` lives in the upstream project service, not this repo: https://github.com/linuxfoundation/lfx-v2-project-service/blob/main/docs/lfid-invite-flow.md. It documents that a settings entry with an empty username triggers the invite email and is promoted in place on acceptance, which is what the Invited / Invite Sent labels key on. I'll make the PR description say so explicitly.

File: apps/lfx-one/src/app/modules/dashboards/components/formation-invite-dialog/formation-invite-dialog.component.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            return throwError(() => error);
+          }
+          // No LF account for this address: re-send WITH the name, which the BFF treats as a
+          // manual add and stores email-only — the shape upstream sends the invite email for.
+          return this.permissionsService
+            .addUserToProject(this.data.projectUid, { name: value.name, email: value.email, role: value.role })
+            .pipe(map((): FormationInviteOutcome => 'invite_sent'));
+        }),
+        take(1),
+        takeUntilDestroyed(this.destroyRef)
```
[the remark this answers is not in the document]
Comment (niravpatel27) [reply apps/lfx-one/src/app/modules/dashboards/components/formation-invite-dialog/formation-invite-dialog.component.ts]: Fixed in 59df0c30ae054a37ce20536bca013a05e1a70df9 — back to `take(1)` only, matching `staff-edit-dialog`. You're right that binding a write to the dialog's lifetime is the wrong trade: aborting an in-flight permissions write on teardown could drop the settings-cache eviction (and the toast) for a write that already landed upstream. The stream completes on its own after one result, so nothing leaks. The comment now records the reason so the next reviewer doesn't re-add it.

**Comment by niravpatel27** [comment]
@dealako one amendment to the list above: the `takeUntilDestroyed` I added on the submit subscription (your minor #6) is reverted in 59df0c30ae054a37ce20536bca013a05e1a70df9. Cursor Bugbot pointed out — correctly — that binding an HTTP *write* to the dialog's lifetime aborts the in-flight request on teardown, which could drop the settings-cache eviction and toast for a write that already reached upstream. It now uses `take(1)` alone, exactly as `staff-edit-dialog` does, with the reason recorded in a comment. The eviction-comment rewording from #6 stands.

**Review by audigregorie — APPROVED** [review]

**Comment by dealako** [comment]
Hey @niravpatel27, thanks for the fast and thorough turnaround on the last round — great to see the writer gate, log masking, and the client-side robustness fixes land together.

**Overall impression:** This is a solid follow-up. The two prior blocking issues (missing writer gate, unmasked email in the add success log) are genuinely fixed and verified server-side by tracing the actual call order and fail-closed behavior, not just re-reading the diff. The client-side hardening (trim-before-validate, empty-email filtering, loading-state guard, `lfx-button` swap, `takeUntilDestroyed`) all check out too. What's left is a cluster of smaller gaps: the demote-guard fix has a narrower blind spot than the original bug, the name-length cap is enforced only in the browser, and two of the new regression tests don't actually exercise the scenario they're meant to protect.

**Nice work:**
- The writer gate in `updateProjectPermissions` (project.service.ts) runs as the very first awaited statement, ahead of the settings read and the directory lookup, and rethrows (fails closed) on an access-service error — that's exactly the pattern `updateProjectStaff` already used, applied correctly to close the oracle.
- Every log call site touched in the add/update/remove chain now goes through `maskIdentifierForLogs`/`maskEmailForLogs` — verified across `project.controller.ts` and `project.service.ts`, not just the one line flagged last round.
- `formation-invite-dialog.component.ts` now patches trimmed name/email back into the form before `markAllAsTouched`, so a pasted address with whitespace normalizes and validates correctly — a real fix, not a workaround.

**Revision tracking:**
- **Resolved** — Missing server-side writer gate / directory oracle: `a24842b` adds `checkSingleAccessStrict({ access: 'writer' })` before any settings read or directory lookup, fails closed on access-check error.
- **Resolved** — Unmasked invitee email in BFF success log: `a24842b` masks the identifier at every log site in the add/update/remove chain.
- **Partially addressed** — Add-is-upsert demote bug: the new 409 `ALREADY_ON_PROJECT` guard correctly blocks the original username-match scenario, but it reuses `matchesUser`, whose email fallback only fires when the existing entry has **no** username. A manual (email-only) add against an already-listed member who *does* have a username still slips past the guard — see Minor findings below. Also, the new 409 broke without adaptation in the other consumer of this endpoint (Settings → Add user), which now shows a generic "Failed to add user" instead of the server's message.
- **Resolved** — Empty-email false-positive duplicate guard: the card now filters empty addresses out of `existingEmails`. (Note: the dialog-side unit test for this doesn't actually exercise the fix — see Minor findings.)
- **Resolved** — Trim before validate: `onSubmit` now patches trimmed values back into the form before validation runs; verified correct for both name and email.
- **Resolved** — Miss-then-failure test: added and passes.
- **Resolved** — Hand-rolled buttons → `lfx-button`: Cancel/Send now use `lfx-button` with `[loading]`, matching `staff-edit-dialog`.
- **Resolved** — Inaccurate DialogService-scope comment / missing `takeUntilDestroyed`: comment reworded accurately, subscription now uses `takeUntilDestroyed`.
- **Partially addressed** — Uncapped name field: a client-side `Validators.maxLength(200)` and native `maxlength` were added, but `addUserToProjectPermissions` on the BFF still performs no length check before persisting `name` and forwarding it to the outbound invite email — the actual trust boundary is unenforced.

**Minor: 7 issues**
- 409 conflict guard misses username-bearing existing members on manual add: `apps/lfx-one/src/server/services/project.service.ts:704-711` — `matchesUser`'s email fallback only fires when the stored entry has no `username`, so an already-listed member with a username can be duplicated (plus a duplicate invite email) via the email-only re-send path instead of being blocked with 409.
- Name-length cap is client-side only: `apps/lfx-one/src/server/controllers/project.controller.ts:279-323` builds `manualUserInfo` from `userData.name` with only a truthiness check; the 200-char cap added in the dialog is bypassable by any direct request.
- New 409 not surfaced in Settings → Add user dialog: `apps/lfx-one/src/app/modules/settings/components/user-form/user-form.component.ts:128-135,225-230` reads `error.error?.message` (a key the envelope never emits) instead of `serverAuthoredMessage`, so re-adding an existing member there now shows a generic, unactionable retry message.
- New 409 test only covers the username-identifier path: `apps/lfx-one/src/server/services/project.service.spec.ts:2387-2396` passes a bare username, so `originalEmail` is `''` and the directory-resolve → email-match branch (the actual path the original bug took) is never exercised.
- Empty-email duplicate-guard spec doesn't test the fix: `apps/lfx-one/src/app/modules/dashboards/components/formation-invite-dialog/formation-invite-dialog.component.spec.ts:128-132` passes identically with or without the `email.length > 0` guard clause, since the fixture's `existingEmails` never contains an empty string.
- Masking is incomplete for manually-added (email-identified) users on the PUT/DELETE routes: the raw address still appears via `req.originalUrl`/`req.path` in request and error logs even though the `username` field in structured logs is now masked — pre-existing gap, but worth closing in the same pass.
- Directory-membership oracle remains for any project writer on `POST /:uid/permissions` (no rate limit) — a writer can still enumerate arbitrary addresses against the LF directory via the 404-vs-201 split. Likely an accepted tradeoff of add-by-email UX, but should be an explicit decision (or rate-limited) rather than assumed.

**Nit: 2 issues**
- A `maxlength` violation on the name field renders the "Enter the person's name." required-error copy/id instead of a length-specific message: `apps/lfx-one/src/app/modules/dashboards/components/formation-invite-dialog/formation-invite-dialog.component.ts:166-175`. Low practical impact since the native `maxlength` attribute already blocks typing/pasting past 200 chars.
- Malformed comment sentence introducing the duplicated writer gate: `apps/lfx-one/src/server/services/project.service.ts:650` — "the gate `updateProjectStaff` runs, for the same reason" reads as a fragment.

**AI bot reconciliation:** Cursor Bugbot posted only a PR-level summary on this commit (no line findings) — nothing to reconcile against. CodeRabbit's auto-review remains disabled on this repo.

Needs changes before approval.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by dealako — CHANGES_REQUESTED** [review]
Summary posted: https://github.com/linuxfoundation/lfx-self-serve/pull/2728#issuecomment-5734848190

Requesting changes for: the 409 duplicate-add guard still has a matching gap that lets an already-listed member (one with a username) be duplicated via the manual/email-only re-send path, the name-length cap added this round is client-side only (BFF still unbounded), the new 409 broke the Settings → Add user flow's error message, two of the new regression tests don't actually exercise the scenario they claim to cover, and a residual directory-membership oracle remains for authenticated writers with no rate limit. See the summary comment for the full findings list.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by dealako** [comment]
Hey @niravpatel27 — thanks for turning this around fast, and for the thorough follow-up on each point from the last round.

**Overall impression:** This revision closes both blocking findings from the previous review cleanly, with tests that actually assert the new behavior rather than just padding coverage. The remaining gaps are narrow: one edge case in the add-by-email duplicate path that self-corrects via the client's existing retry, a missing regression test for the exact scenario the last commit fixes, and a small UX/validation-hardening nit. Nothing here changes the shape of the feature.

**Nice work:** the writer-gate fix closes the 404-vs-403 oracle at the right layer — `checkSingleAccessStrict` now runs as step 0 in `updateProjectPermissions`, before any settings read or NATS lookup, and `project.service.spec.ts` proves it with an assertion that the downstream calls are never reached on a denied check. The log-masking sweep was also thorough — every log site across the add/update/remove permission handlers was updated, not just the one flagged.

**Revision tracking:**
- **Resolved** — Missing server-side writer gate creating a directory-membership oracle: `checkSingleAccessStrict` now gates `updateProjectPermissions` before the directory lookup, fail-closed on an access-check error. (https://github.com/linuxfoundation/lfx-self-serve/commit/a24842bc92b7ea48c3d51f3542b60dbee679547d)
- **Resolved** — Unmasked invitee email in the BFF success log: all identifier logging in `project.controller.ts`/`project.service.ts` now goes through `maskIdentifierForLogs`/`maskEmailForLogs`. (https://github.com/linuxfoundation/lfx-self-serve/commit/a24842bc92b7ea48c3d51f3542b60dbee679547d)
- **Partially addressed** — Duplicate-guard/role-overwrite correctness: the username-matched re-add case now correctly returns 409 `ALREADY_ON_PROJECT` instead of silently re-filing the role, and the empty-string false-duplicate case is fixed. However, an add-by-bare-email request against someone previously added *manually* (no directory username) still throws a directory-miss `ResourceNotFoundError` before the 409 check is reached, because `resolveEmailToUsername` runs first for that path — see Minor finding below.
- **Partially addressed** — Test-coverage gaps: strong new coverage was added for the writer gate, 409 refusal, masking, and dialog validation, but no test exercises the specific regression the final commit (59df0c3) fixes — an in-flight write surviving component teardown under `take(1)`-only.

Blocking: 0 issues

Minor: 3 issues
- Add-by-email duplicate check bypassed on first attempt for a manually-added person: `apps/lfx-one/src/server/services/project.service.ts:690-720`: the `add`+bare-email path calls `resolveEmailToUsername` before the new 409 check runs, so a duplicate manually-added user gets a directory-miss 404 instead of 409 on the first request (it only self-corrects via the invite dialog's built-in retry-with-name, so end-state is fine for this UI, but any other caller sees the wrong error).
- No test covers the take(1)-teardown fix the last commit makes: `apps/lfx-one/src/app/modules/dashboards/components/formation-invite-dialog/formation-invite-dialog.component.spec.ts`: no spec destroys the component mid-submit and asserts the write still completes and `invalidateProjectSettings` still fires — exactly the regression 59df0c3 exists to prevent.
- Name-field maxlength error renders the wrong message: `apps/lfx-one/src/app/modules/dashboards/components/formation-invite-dialog/formation-invite-dialog.component.ts:169-178` (+ `.html:21-23`): hitting the new 200-char cap shows "Enter the person's name." instead of a length-specific message, since `initNameErrorId()` doesn't branch on `hasError('maxlength')`.

Nit: 2 issues
- Name length cap is client-side only: `packages/shared/src/constants/formation-people.constants.ts` / `apps/lfx-one/src/server/controllers/project.controller.ts:314-323` / `apps/lfx-one/src/server/services/project.service.ts:739-741`: the server never validates `userData.name` length before persisting/emailing it, so a direct API caller can bypass the 200-char guarantee the constant documents.
- Class-level doc comment not updated: `apps/lfx-one/src/app/modules/dashboards/components/formation-invite-dialog/formation-invite-dialog.component.ts:33-44`: doesn't mention the new name-length cap or the empty-string-is-never-a-duplicate carve-out.

**AI bot reconciliation:** Agree with Copilot's whitespace-validation finding being resolved — trimming landed in 3c1297f. Agree Cursor Bugbot's flagged issue (submit subscription bound to dialog lifetime, dropping in-flight requests on teardown) is fixed in 59df0c3 by reverting to `take(1)`; note that fix currently ships without regression coverage (see Minor finding above).

Needs changes before approval.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by dealako — CHANGES_REQUESTED** [review]
Follow-up review: both prior blocking findings (writer-gate oracle, unmasked log) are resolved with good test coverage. Requesting changes for the remaining gaps: an add-by-email duplicate check that's bypassed on the first attempt for a manually-added user (self-corrects via the dialog's retry, but wrong for any other caller), a missing regression test for the take(1)-teardown fix, a mislabeled maxlength error message, and 3 total minor issues exceeds the 2-minor approval threshold. Full findings in the summary comment: https://github.com/linuxfoundation/lfx-self-serve/pull/2728#issuecomment-5734859041

---
_Generated by [Claude Code](https://claude.ai/code)_
